mt_gateway_test.go

```diff
@@ -0,0 +1,369 @@
+	require.NoError(t, os.Setenv("RSERVER_BACKEND_CONFIG_USE_HOSTED_BACKEND_CONFIG", "false"))
+	require.NoError(t, os.Setenv("RUDDER_TMPDIR", rudderTmpDir))
+	require.NoError(t, os.Setenv("HOSTED_MULTITENANT_SERVICE_SECRET", "so-secret"))
+	require.NoError(t, os.Setenv("DEPLOYMENT_TYPE", string(deployment.MultiTenantType)))
+	if testing.Verbose() {
+		require.NoError(t, os.Setenv("LOG_LEVEL", "DEBUG"))
+	}
+
+	// The Gateway will not become healthy until we trigger a valid configuration via ETCD
+	// TODO: this is to be reviewed after "Review health checkpoint (probes)
+	// https://www.notion.so/rudderstacks/Review-health-checkpoint-probes-ec33b45c1b7541f3bf802f3276667920
+	etcdReqKey := getGatewayWorkspacesReqKey(t, serverInstanceID)
+	_, err = etcdContainer.Client.Put(ctx, etcdReqKey, `{"workspaces":"`+workspaceID+`","ack_key":"test-ack/1"}`)
+	require.NoError(t, err)
+
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		Run(ctx)
+	}()
+
+	serviceHealthEndpoint := fmt.Sprintf("http://localhost:%d/health", httpPort)
+	t.Log("serviceHealthEndpoint", serviceHealthEndpoint)
+	health.WaitUntilReady(ctx, t,
+		serviceHealthEndpoint,
+		time.Minute,
+		250*time.Millisecond,
+		"serviceHealthEndpoint",
+	)
+
+	select {
+	case ack := <-etcdContainer.Client.Watch(ctx, "test-ack/1"):
+		v, err := unmarshalWorkspaceAckValue(t, &ack)
+		require.NoError(t, err)
+		require.Equal(t, "RELOADED", v.Status)
+		require.Equal(t, "", v.Error)
+	case <-time.After(20 * time.Second):
+		t.Fatal("Timeout waiting for test-ack/1")
+	}
+
+	require.Empty(t, webhook.Requests(), "webhook should have no requests before sending the events")
+	sendEventsToGateway(t, httpPort, writeKey)
+
+	var (
+		eventPayload string
+		message      map[string]interface{}
+	)
+	require.Eventually(t, func() bool {
+		return postgresContainer.DB.QueryRowContext(ctx,
+			"SELECT event_payload FROM gw_jobs_1 WHERE workspace_id = $1", workspaceID,
+		).Scan(&eventPayload) == nil
+	}, time.Minute, 50*time.Millisecond)
+	require.NoError(t, json.Unmarshal([]byte(eventPayload), &message))
+
+	batch, ok := message["batch"].([]interface{})
+	require.True(t, ok)
+	require.Len(t, batch, 1)
+	require.Equal(t, message["writeKey"], writeKey)
+	for _, msg := range batch {
+		m, ok := msg.(map[string]interface{})
+		require.True(t, ok)
+		require.Equal(t, "anonymousId_1", m["anonymousId"])
+		require.Equal(t, "identified_user_id", m["userId"])
+		require.Equal(t, "identify", m["type"])
+		require.Equal(t, "1", m["eventOrderNo"])
+		require.Equal(t, "messageId_1", m["messageId"])
+	}
+
+	// Only the Gateway is running, so we don't expect any destinations to be hit.
+	require.EqualValues(t, 0, webhook.RequestsCount(), "webhook should have no requests because there is no processor")
+
+	serverModeReqKey := getETCDServerModeReqKey(t, serverInstanceID)
+	t.Logf("Server mode ETCD key: %s", serverModeReqKey)
+
+	{ // Trigger degraded mode, the Gateway should still work
```

Comment:
is there a way to split this test into sub-tests and at the same time make sure that sub-tests will run in order and not in parallel?
E.g. something similar to Ginkgo's [Ordered](https://onsi.github.io/ginkgo/#ordered-containers)

Reply:
I can do a `t.Run()`, they will run sequentially. I didn't do it because I prefer not to give the impression that you could run them separately. For example the main test ends because we fire an invalid workspace change. I can move it in `t.Run()` but then I'll have to add a `cancel()` to the main test in case you don't run the "invalid workspace change" scenario.

It feels to me that given that it's just doing 3 things it's easier to just follow them as is. It makes the main test a bit longer but easier to follow since it's clear that we're just sequentially changing state and not resetting anything between tests.

So yeah, I could split it but I don't think it'll make our life easier :thinking:

Reply:
How do multiple `t.Run` on the same test behave with ginkgo's -parallel flag?

Reply:
> We run tests with ginkgo. How do multiple `t.Run` on the same test behave with ginkgo's -parallel flag?

Do you mean with `--procs`? If you don't put `t.Parallel()` they are not executed in parallel.

```go
package ciao

import "testing"

func TestSomething(t *testing.T) {
	var count int
	t.Run("aris-1", func(t *testing.T) {
		t.Parallel()
		count++
	})
	t.Run("aris-2", func(t *testing.T) {
		t.Parallel()
		count++
	})
}
```

Run the above with `ginkgo --procs=24 --race ./ciao` and you get a data race warning. Do the same without the two `t.Parallel()` and it will be fine :+1: 

Btw, I will at least remove the extra scope: https://deepsource.io/gh/rudderlabs/rudder-server/run/ed1f1626-8c90-4ceb-a79a-1ba445e73774/go/CRT-A0008 :man_facepalming:

Reply:
thanks for taking the time to verify :) I would keep the extra scope and add a `// skipcq: CRT-A0008` instead, it is useful, to me!

Reply:
Did some reorganization, I think it's better now :+1: 
The tests can only be fired singularly, the only thing we can't do is move the last one up since it'll cause a termination so we have to keep it at the bottom (also no parallelism allowed obviously).